### PR TITLE
Compatibility with database_cleaner v2 adapter gems

### PR DESCRIPTION
**Pull request by botandrose**
## Summary

Hi folks! This PR aims to resolve some incompatibilities with cucumber-rails and the upcoming major v2 release of database_cleaner.

## Details
In order to get access to the database_cleaner configuration in a ORM-agnostic way, we first try to require `database_cleaner/core` (which is only available in the database_cleaner v2 family of gems), then fall back to requiring `database_cleaner` for folks still on v1, finally falling back to silence if neither is available, since database_cleaner is an optional dependency.

## Motivation and Context

Last week, we cut beta versions of the next major version of database_cleaner. The main change here is that all of the ORM adapters have been split off into their own gems, e.g. database_cleaner-active_record, database_cleaner-redis, etc. Understanding that the majority of database_cleaner's current users are using it with Rails/ActiveRecord (and this gem, too), we chose to make the database_cleaner v2.0.0 gem a metagem that depends on database_cleaner-active_record, and simply loads that when `require "database_cleaner"` is run. This means that the vast majority of our users won't have to touch a thing to upgrade from database_cleaner v1 to v2. 

However, not everyone is only using ActiveRecord. There are those who can't get away with just leaving `gem "database_cleaner"` as-is in their Gemfile, and they have a bit more work to do. For example, let's say you're using Rails with Sequel and Redis, and want to upgrade to database_cleaner v2.0.0. No big deal, you just replace `gem "database_cleaner"` with `gem "database_cleaner-sequel"` and `gem "database_cleaner-redis"` and you're good to go. However, in this situation, the current implementation of cucumber-rails' database_cleaner integration will silently fail.

The underlying issue is that none of the ORM adapters have a `lib/database_cleaner.rb` file (only the database_cleaner metagem has it, for backwards compatibility with the common case), so the `require "database_cleaner"` call in the "lib/cucumber/rails/hooks/database_cleaner.rb" will silently fail since `LoadError` is being caught and discarded, the only visible clue being that the database state is now persisting between test cases. Personally, I consider this issue to be a blocker for releasing final v2.0.0 versions of the database_cleaner family of gems.

## How Has This Been Tested?

The first commit in this PR adds a new cucumber feature that fails, exposing this issue. It does this by simply adding `gem "database_cleaner-active_record"` to the Gemfile, instead of `gem "database_cleaner"`. This was the simplest way to expose the issue without adding a dev dependency on another ORM. This is also a valid and supported way to use database_cleaner with ActiveRecord, BTW.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

This change should be transparent to consumers of the library.

### CODE REVIEW

**Comment by botandrose** [comment]
Okay, tests have all completed for the failing test commit. I've checked the output for all the runners, and they are all failing as expected, with the following:

```
Using the default profile...
Feature: Create widgets
  Background: 2 initial widgets # features/widgets.feature:2
    Given I have 2 widgets      # features/step_definitions/widget_steps.rb:1
  Scenario: Add 3 widgets        # features/widgets.feature:5
    When I create 3 more widgets # features/step_definitions/widget_steps.rb:7
    Then I should have 5 widgets # features/step_definitions/widget_steps.rb:13
  Scenario: Add 7 widgets        # features/widgets.feature:9
    When I create 7 more widgets # features/step_definitions/widget_steps.rb:7
    Then I should have 9 widgets # features/step_definitions/widget_steps.rb:13
      expected: 9
           got: 14
      (compared using ==)
       (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/widget_steps.rb:14:in `"I should have {int} widgets"'
      features/widgets.feature:11:in `I should have 9 widgets'
Failing Scenarios:
cucumber features/widgets.feature:9 # Scenario: Add 7 widgets
2 scenarios (1 failed, 1 passed)
6 steps (1 failed, 5 passed)
0m0.155s
..................................................................
(::) failed steps (::)
expected that command "bundle exec rake cucumber" has exit status of "0", but has "1". (RSpec::Expectations::ExpectationNotMetError)
./features/step_definitions/cucumber_rails_steps.rb:72:in `"the feature run should pass with:"'
features/database_cleaner.feature:83:in `the feature run should pass with:'
Failing Scenarios:
cucumber features/database_cleaner.feature:46 # Scenario: Create records in background with database_cleaner-active_record
```

**Comment by botandrose** [comment]
Okay, we're green! Ready for review. What do y'all think?

**Comment by luke-hill** [comment]
Hi there. So provisionally I'm ok with this. I'm trying to find a commit, but maybe you can help.

Someone mentioned that with a new-er version of Rails (I can't remember if it was 5.2 or 6.0), database cleaner was no longer required and sort of auto included. But I can't remember the details as it was a long time ago. I remember the thought process here was we were going to *remove* database cleaner at some point.

If I'm wrong that's fine. You may be able to piece things together better for me. I'll try have a dig around. For now can we just hold off. But if everything's ok I'm happy with the approach here, just need to confirm my original thought process was wrong.

**Comment by luke-hill** [comment]
https://github.com/cucumber/cucumber-rails/issues/406 &
https://github.com/cucumber/cucumber-rails/issues/459

**Comment by BrianHawley** [comment]
It was in Rails 5.1, and it's not strictly that database_cleaner isn't required. It was more that they made it possible to run tests in a mode where all threads share a database connection, and this mode makes it possible to have transaction-mode cleanup actually work in multi-threading environments like capybara.

However, that mode does change the behavior of the app: all threads are sharing one database connection. Rails doesn't have built-in support for any of the other database_cleaner cleanup modes, which you'd still need to use when testing certain kinds of concurrency issues.

Fortunately the new test concurrency mode works with database_cleaner's transactional cleanup mode, so even if you have it installed, it can still be used.

**Comment by luke-hill** [comment]
Ok so in essence, adding this new additional permutation won't alter the use case for people who no longer want to use database cleaner in rails 5.1+ ?

File: lib/cucumber/rails/hooks/database_cleaner.rb [hunk 1]
```diff
@@ -1,15 +1,21 @@
 # frozen_string_literal: true
 
 begin
-  require 'database_cleaner'
+  require 'database_cleaner/core'
+rescue LoadError
+  begin
+    require 'database_cleaner'
+  rescue LoadError
+    Cucumber.logger.debug('database_cleaner not present.')
```
Comment (luke-hill) [inline lib/cucumber/rails/hooks/database_cleaner.rb:9]: Probably worth hinting that this has checked both gems right?

**Comment by luke-hill** [comment]
Also when you come back to this, could you rebase with latest master, and add a changelog entry at the top in unreleased

File: lib/cucumber/rails/hooks/database_cleaner.rb (hunk 1 above)
Comment (botandrose) [reply lib/cucumber/rails/hooks/database_cleaner.rb:9]: Sure! Are you imagining something like "neither database_cleaner v1 or v2 present"?

**Comment by botandrose** [comment]
@luke-hill Rebased and changelog updated. Sorry it took so long to get back to this. Last week was crazy IRL. Looks like I missed the v2.1 release train, too. Crap!

**Comment by luke-hill** [comment]
Trying to see if the failure was a genuine timeout or something else (Fairly sure travis just had a moment).

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,15 +5,16 @@ on how to contribute to Cucumber.
...
+ *
 
 ### Changed
 
  *
 
 ### Fixed
 
- *
+ * Database cleaning silently fails when using database_cleaner v2 adapter gems
```
Comment (luke-hill) [inline CHANGELOG.md:16]: should this say something like "no longer fails", or have I misunderstood this?

File: features/support/cucumber_rails_helper.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@ def install_cucumber_rails(*options)
     add_gem 'selenium-webdriver', '~> 3.11', group: :test
     add_gem 'rspec-expectations', '~> 3.7', group: :test
     add_gem 'database_cleaner', '>= 1.8.0', group: :test unless options.include?(:no_database_cleaner)
+    add_gem 'database_cleaner-active_record', '>= 2.0.0.beta2', group: :test if options.include?(:database_cleaner_active_record)
```
Comment (luke-hill) [inline features/support/cucumber_rails_helper.rb:16]: Is this beta copy the best we have (This helper is used for everyone who uses the gem). Do we have any idea how close to a stable 2.x revision we are?

File: lib/cucumber/rails/hooks/database_cleaner.rb (hunk 1 above)
Comment (luke-hill) [reply lib/cucumber/rails/hooks/database_cleaner.rb:9]: Something like that yeh

File: CHANGELOG.md (hunk 2 above)
Comment (botandrose) [reply CHANGELOG.md:16]: Ah, I think I see. I was describing the bug, and you would prefer that I describe the bugfix?

File: features/support/cucumber_rails_helper.rb (hunk 3 above)
Comment (botandrose) [reply features/support/cucumber_rails_helper.rb:16]: Yes, this issue is the last blocker for releasing v2.0.0 final. So, as soon as this PR is merged, and out in the wild in a cucumber-rails gem release, I will cut the final v2.0.0 release of database_cleaner and its family of adapter gems.

**Comment by botandrose** [comment]
@luke-hill I have just pushed the requested changes the changelog message and the warning message.

File: features/support/cucumber_rails_helper.rb (hunk 3 above)
Comment (botandrose) [reply features/support/cucumber_rails_helper.rb:16]: If we could coordinate a simultaneous release, we could change this to `>= 2.0.0`

Comment (deivid-rodriguez) [reply features/support/cucumber_rails_helper.rb:16]: I'm confused, how is this helper used by everyone using this gem? This file is shipped with the gem, but its folder (`features`) is not exposed as a $LOAD_PATH directory. I thought these were just internal helpers for testing `cucumber-rails`.

**Comment by aslakhellesoy** [comment]
Hi @botandrose,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Pull request merged**